protomap/proto.go

```diff
@@ -704,6 +729,153 @@ func protoFromPathsInternal(p proto.Message, vals map[*gpb.Path]any, valPrefix,
+		}
+
+		// Find the parts of the path that are not the list -- we assume that this is 2 elements since
+		// we are in a compressed schema.
+		// TODO(robjs): Currently, this may report incorrectly in an uncompressed schema, but we don't have
+		// a signal to indicate this. One needs to be added to the generated protobufs.
+		keyPath := &gpb.Path{
+			Elem: append(append([]*gpb.PathElem{}, protoPrefix.Elem...), absPath.Elem[len(protoPrefix.Elem):len(protoPrefix.Elem)+2]...),
+		}
+		var alreadySeen bool
+		for _, ek := range keys {
+			if reflect.DeepEqual(ek, k.Key) {
+				alreadySeen = true
+				break
+			}
+		}
+		if !alreadySeen {
+			keys = append(keys, k.Key)
+			keyPaths = append(keyPaths, keyPath)
+		}
+	}
+
+	le := m.ProtoReflect().NewField(fd).List()
+	for i, key := range keys {
+		listElemChildren, err := findChildren(vals, valPrefix, keyPaths[i], false, false)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("logic error, error returned from extracting list member children, %v", err)
+		}
+
+		// We now need to create the "XXXKey" message, and populate the key values, subsequent values are then populated
+		// into the one protobuf message field.
+		childMsgEmpty := le.NewElement().Message()
+		childMsgTarget := le.NewElement().Message()
+
+		// Walk through the fields of the XXXKey message that we just created. We use the childMsgEmpty here so that we
+		// don't change the message whilst iterating which causes us to revisit that field. We set the values in
+		// the childMsgTarget message.
+		var retErr error
+
+		// Store the key values that we received, to make sure that they are mapped during iteration
+		// through the protobuf fields.
+		remainingKeys := map[string]bool{}
+		for n := range key {
+			remainingKeys[n] = true
+		}
+		unpopRange{childMsgEmpty}.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+			// We have one field that is a message in a key message, which is the payload. The remaining fields are the keys.
+			switch fd.Kind() {
+			case protoreflect.MessageKind:
+				m := childMsgTarget.NewField(fd).Message()
+				// We must ignore extra fields from this point in the recursion, because keys map to fields that
+				// are not present in the generated protobuf.
+				if err := protoFromPathsInternal(m.Interface(), listElemChildren, keyPaths[i], keyPaths[i], true); err != nil {
+					retErr = err
+					return false
+				}
+				childMsgTarget.Set(fd, protoreflect.ValueOfMessage(m))
+			default:
+				paths, err := annotatedSchemaPath(fd)
+				if err != nil {
+					retErr = err
+					return false
+				}
+
+				var keyName string
+				for _, p := range paths {
+					keyName = p.Elem[len(p.Elem)-1].Name
+					break
+				}
+				if key[keyName] == "" {
+					retErr = fmt.Errorf("field %s, missing key %s, got keys: %v", fd.FullName(), keyName, key)
+					return false
+				}
```

Comment:
Should this be checking `ok` rather than "" since, perhaps, "" could be a valid key in rare situations?

Reply:
Good catch, yes, I guess unlikely that this is the case, but possible. Fixed.